lib-intro.tex

```diff
@@ -72,3 +72,124 @@
+signature of the function or functions in namespace
+tcode{std}, ensuring that overload resolution will fail
+unless a declaration of the signature is found that is either more
+specialized~(\cxxref{temp.func.order}) or more
+constrained~(N4377 [temp.constr.order]) than the poison pill.
+
+\enterexample \tcode{ranges::swap}~(\ref{utility.swap}) uses
+the name \tcode{swap} which has two overloads defined in
+namespace \tcode{std}~(C++ \cxxref{utility.swap}):
+\begin{codeblock}
+template <class T>
+void swap(T&, T&) noexcept(is_nothrow_move_constructible<T>::value &&
+                           is_nothrow_move_assignable<T>::value);
+template <class T, size_t N>
+void swap(T(&a)[N], T(&b)[N]) noexcept(noexcept(swap(*a, *b)));
+\end{codeblock}
+\tcode{ranges::swap} specifies that the overload resolution
+it performs internally for \tcode{swap(E1, E2)} for some
+expressions \tcode{E1} and \tcode{E2} is performed in a
+context that includes the declarations
+\begin{codeblock}
+template <class T>
+void swap(T&, T&) = delete;
+template <class T, size_t N>
+void swap(T(&)[N], T(&)[N]) = delete;
+\end{codeblock}
+which ensures that unqualified name lookup~(\cxxref{basic.lookup.unqual})
+will never reach the definitions in namespace \tcode{std}, and that
+overload resolution will fail due to ambiguity if the signatures
+in namespace \tcode{std} are found by argument dependent lookup.
+\exitexample
+
+This technique is not without consequences to users of the library
+specified in this document: user-defined overloads of such signatures
+intended to be found by argument dependent lookup must be more
+specialized or more constrained than the ``poison pill'' declarations.
+\enterexample
+\begin{codeblock}
+namespace X {
+struct A {};
+void swap(A&, A&);           // \#1
+}
+
+namespace Y {
+template <class T> struct B {};
+template <class T, class U>
+void swap(B<T>&, B<U>&);     // \#2
+}
+
+namespace Z {
+template <class T> struct D : std::string {};
+template <class T>
+  requires 6 * 7 == 42
+void swap(T&, T&);           // \#3
+}
+\end{codeblock}
+Calls to \tcode{ranges::swap} with two lvalues of type \tcode{A}
+result in \#1 being called, since it is more specialized than the
+``poison pill'' overload \tcode{swap(T\&, T\&)}. A call with lvalues
+of type \tcode{B<int>} and \tcode{B<double>} will invoke \#2, since
+the poison pill overload is not viable. A call with two lvalues of
+type \tcode{B<int>}, however, will fail overload resolution since
+both \#2 and the poison pill overload are viable and neither is
+more specialized nor more constrained.
+
+A call with lvalues of type \tcode{D<int>} and \tcode{D<int>}
+results in a lookup set containing \#3, the poisin pill overload,
+and \tcode{std::swap(T\&, T\&)} since \tcode{D} inherits from
+\tcode{std::string} making \tcode{std} an associated namespace.
+Overload resolution will unambiguously select \#3 since it
+is more constrained than the other two overloads.
+\exitexample
+}
```

Comment:
When I started thinking about introducing "poison pill" as a term-of-art, I noticed that this "example" is the majority of the section and I knew I had gone off the rails. I want to make two points here: (1) the deleted overloads prevent calling `std` functions, and (2) user overloads necessarily must be more specialized/more constrained than the deleted overloads to hook the customization point when the deleted overloads are viable. (2) has a subtle danger:

```c++
template <class T> struct A {};
template <class T, class U> void swap(A<T>&, A<U>&) {}
```
If I call `ranges::swap` with lvalues `A<int>` and `A<double>`, my swap overload is invoked. If I call with `A<int>` and `A<int>` it is not invoked, and I don't get a compile error - `Movable<A>()` is satisfied so the default swap implementation is used.

How about:
```tex
\enternote Many of the customization points objects in the library
evaluate function call expressions with an unqualified name which
results in a call to a user-defined function found by argument
dependent name lookup~(\cxxref{basic.lookup.argdep}). To preclude
such an expression resulting in a call to unconstrained functions
with the same name in namespace \tcode{std}, customization point
objects specify that lookup for these expressions is performed in
a context that includes deleted overloads matching the signatures
of overloads defined in namespace \tcode{std}. When the deleted
overloads are viable, user-defined overloads must be more
specialized~(\cxxref{temp.func.order}) or more
constrained~(N4377 [temp.constr.order])  to be used by a
customization point object. \exitnote
```